Rewrite PathFilter

JoeLametta opened the pull request:
Added filter options:
- dot (replace leading dot with _)
- posix (replace illegal chars in *nix OSes with _)
- vfat (replace illegal chars in VFAT filesystems with _)
- whitespace (replace all whitespace chars with _)
- printable (replace all non printable ASCII chars with _)

Removed filter options:
- fat (replaced with vfat)
- special

Fixes #313.

Freso commented on whipper/test/test_common_path.py:
```diff
@@ -9,22 +9,28 @@
 
     def setUp(self):
-        self._filter = path.PathFilter(special=True)
+        self._filter_default = path.PathFilter()
```
Isn't this the same as `self._filter_posix`? The default is "`posix=True, fat=False`", no? Maybe it'd be better to have a test `self._filter_none = path.PathFilter(posix=False, fat=False)`?

Freso commented on whipper/test/test_common_path.py:
```diff
@@ -9,22 +9,28 @@
+        self.assertEqual(self._filter_fat.filter(part), u'A Word_ F__k you_')
 
-    def testSpecial(self):
-        part = u'<<< $&*!\' "()`{}[]spaceship>>>'
```
This seems like a good string to test still. Any particular reason you're not including this anymore?

Freso commented:
Also, it'd be nice with in-line comments for sources for the "bad characters". Esp. the POSIX ones, since I'm pretty sure it's a file system specific thing and not necessarily a POSIX standard thing? I might be wrong though.

Freso commented on whipper/common/path.py:
```diff
@@ -26,42 +26,18 @@ class PathFilter(object):
-                          '_', path, re.UNICODE)
-
+        if self._posix:
+            path = re.sub(r'[\/\x00]', '_', path, re.UNICODE)
```
I know this raises complexity, but should we also check for (and replace) `.` if it's the first character? For most (all?) POSIX systems `.*` are hidden files, and we probably don't want to "hide" ripped files.

There's [only one artist](https://musicbrainz.org/search?query=artist%3A%22.%22&type=artist&limit=25&method=advanced) but [a lot of recordings/tracks](https://musicbrainz.org/search?query=recording%3A%22.%22&type=recording&limit=25&method=advanced) (in MusicBrainz) with a "." as their first character, so while it's a small chance for someone to encounter this, it is a real case that could be encountered, even using the default settings (`DEFAULT_DISC_TEMPLATE = u'%r/%A - %d/%A - %d'` would result in `ep/. - IHI/. - IHI` for [.’s "IHI"](https://musicbrainz.org/release/fba611e2-2e17-4cae-83c7-3751f6cf734)). Granted, that person has only released digital stuff so far, but someone might want to put the release name first or do something else.

Freso replied:
Oh, https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations for 'POSIX "Fully portable filenames"' also says:
> hyphen must not be first character

So if we need to check for hyphen as first character, we might as well also check for period?

JoeLametta replied:
> I know this raises complexity, but should we also check for (and replace) `.` if it's the first character? For most (all?) POSIX systems `.*` are hidden files, and we probably don't want to "hide" ripped files.

Done.

> hyphen must not be first character

I think this one is no longer relevant.

JoeLametta commented:
> Right now you're testing different strings for different functions. I think it'd make more sense to come up with a selection of strings to test and then test that set of strings for the entire matrix to make sure that they all behave correctly in all instances.
> 
> https://github.com/minimaxir/big-list-of-naughty-strings may be a good place to source some strings for testing from too.

Yeah, need to do this.

I've just updated this pull request: it's still a WIP.

JoeLametta commented:
I've just added another filter: `printable` (disabled by default). It replaces every non-printable ASCII character.

> Right now you're testing different strings for different functions. I think it'd make more sense to come up with a selection of strings to test and then test that set of strings for the entire matrix to make sure that they all behave correctly in all instances.
> 
> https://github.com/minimaxir/big-list-of-naughty-strings may be a good place to source some strings for testing from too.

I've extracted/modified the following strings from that resource (everything is base64 encoded to prevent file encoding issues). What do you think?

```
# ASCII punctuation
ASCII_P = 'LC4vOydbXVwtPTw+Pzoie318XyshQCMkJV4mKjp3cSgpYH4='

# Non-whitespace C0 controls
CTRL_0 = 'AQIDBAUGBwgODxAREhMUFRYXGBkaGxwdHh9/'

# Non-whitespace C1 controls
CTRL_1 = ('woDCgcKCwoPChMKGwofCiMKJworCi8KMwo3CjsKPwpDCkcKSw'
          'pPClMKVwpbCl8KYwpnCmsKbwpzCncKewp8=')

# Whitespace: all of the characters with category Zs, Zl, or Zp (in Unicode
# version 8.0.0), plus U+0009 (HT), U+000B (VT), U+000C (FF), U+0085 (NEL),
# and U+200B (ZERO WIDTH SPACE), which are in the C categories but are often
# treated as whitespace in some contexts.
WHITE = ('CwwgAAoNwoXCoOGagOKAgOKAgeKAguKAg+KAhOKAheKAhuKAh
         '+KAiOKAieKAiuKAi+KAqOKAqeKAr+KBn+OAgA==')

# Unicode additional control characters: all of the characters with
# general category Cf (in Unicode 8.0.0).
CTRL_3 = ('wq3YgNiB2ILYg9iE2IXYnNud3I/hoI7igIvigIzigI3igI7ig'
          'I/igKrigKvigKzigK3igK7igaDigaHigaLigaPigaTigabiga'
          'figajiganigarigavigaziga3iga7iga/vu7/vv7nvv7rvv7v'
          'wkYK98JuyoPCbsqHwm7Ki8Juyo/CdhbPwnYW08J2FtfCdhbbw'
          'nYW38J2FuPCdhbnwnYW686CAgfOggKDzoICh86CAovOggKPzo'
          'ICk86CApfOggKbzoICn86CAqPOggKnzoICq86CAq/OggKzzoI'
          'Ct86CArvOggK/zoICw86CAsfOggLLzoICz86CAtPOggLXzoIC'
          '286CAt/OggLjzoIC586CAuvOggLvzoIC886CAvfOggL7zoIC/'
          '86CBgPOggYHzoIGC86CBg/OggYTzoIGF86CBhvOggYfzoIGI8'
          '6CBifOggYrzoIGL86CBjPOggY3zoIGO86CBj/OggZDzoIGR86'
          'CBkvOggZPzoIGU86CBlfOggZbzoIGX86CBmPOggZnzoIGa86C'
          'Bm/OggZzzoIGd86CBnvOggZ/zoIGg86CBofOggaLzoIGj86CB'
          'pPOggaXzoIGm86CBp/OggajzoIGp86CBqvOggavzoIGs86CBr'
          'fOgga7zoIGv86CBsPOggbHzoIGy86CBs/OggbTzoIG186CBtv'
          'OggbfzoIG486CBufOggbrzoIG786CBvPOggb3zoIG+86CBvw==')

# "Byte order marks", U+FEFF and U+FFFE
BOM = '77u/77++'

# Unicode Symbols
U_SYM = ('xZPiiJHCtMKu4oCgwqXCqMuGw7jPgOKAnOKAmMOFw43DjsOPy'
         '53Dk8OU76O/w5LDmsOG4piDxZLigJ7CtOKAsMuHw4HCqMuGw5'
         'jiiI/igJ3igJlg4oGE4oKs4oC54oC676yB76yC4oChwrDCt+K'
         'AmuKAlMKx')

# Special Unicode Characters Union
#
# A super string recommended by VMware Inc. Globalization Team:
# can effectively cause rendering issues or character-length
# issues to validate product globalization readiness.
UNION = '6KGo44Od44GCQem3l8WSw6nvvKLpgI3DnMOfwqrEhcOx5LiC45CA8KCAgA=='

# Japanese Emoticon
JPN = 'KCDNocKwIM2cypYgzaHCsCk='

# Emoji
EMOJI = '8J+RviDwn5mHIPCfkoEg8J+ZhSDwn5mGIPCfmYsg8J+ZjiDwn5mN'

# Regional Indicator Symbols
RIS = 8J+HuvCfh7jwn4e38J+HuvCfh7gg8J+HpvCfh6vwn4em8J+HsvCfh7g=

# Right-To-Left String
RTL = '15TWuNeZ1rDXqta415R0ZXN02KfZhNi12YHYrdin2Kog2KfZhNiq2ZHYrdmI2YQ='

# Zalgo Text
ZALGO = ('WsyuzJ7MoM2ZzZTNheG4gMyXzJ7NiMy7zJfhuLbNmc2OzK/Mu'
         'cyezZNHzLtPzK3Ml8yu')

# Unicode Upsidedown
U_UPS = 'MDDLmcaWJC0='

# Unicode font
U_FONT = ('77y0772I772FIO+9ke+9le+9ie+9g++9iyDvvYLvvZLvvY/vvZfvvY4g772'
          'G772P772YIO+9iu+9le+9je+9kO+9kyDvvY/vvZbvvYXvvZIg772U772I77'
          '2FIO+9jO+9ge+9mu+9mSDvvYTvvY/vvYc=')
```

There's also this one but I don't think whipper ever generates filenames without any extension.

```
# MSDOS/Windows Special Filenames
S_NAMES = {'CON', 'PRN', 'AUX', 'CLOCK$', 'NUL', 'A:', 'ZZ:',
           'COM1', 'LPT1', 'LPT2', 'LPT3', 'COM2', 'COM3', 'COM4'}
```

JoeLametta commented:
Rebased (fixed merge conflicts).

JoeLametta commented:
Fixed merge conflict.

JoeLametta commented:
@Freso I haven't ignored your comment, just didn't want to let this one die. I hope it will be improved later.
